### ISSUE TITLE
Use tokio instead of custom event loop

### ISSUE THREAD

**Issue by MOZGIII** [RustAudio/cpal#257]
Current event loop implementation works, but it would be great to adopt tokio as it's a wide-spread common API that many thing plug in to.
It seems like the current implementation is compatible with tokio.

**Comment by mitchmindtree** [comment]
Once upon a time cpal had it's own `Futures` based event loop implementation, but I believe it was dropped as 1. it introduced a lot of unnecessary complexity and 2. futures (0.1 at the time) was quite unstable and there was a lot of uncertainty about the design direction.

All that said, I think it's a good idea to revisit the possibility of having a futures based API around the event loop (or at least something that is compatible), but I personally would be happy to wait until the futures module finally lands in std before fully revisiting the idea.

Re tokio - I'm curious what benefits you feel we would gain from using tokio specifically? Off the top of my head, I can't think of what benefit we would gain by using tokio that we wouldn't get from using futures directly. Do you have an example or code snippet in mind of how the API might be improved by leveraging tokio?

**Comment by MOZGIII** [comment]
Sounds good!

My use case is I just want to grab the audio input and send it over the network. On the other end the process woud reverse. With tokio integration, it seems like I could use just a signle thread to do all of it. Although, I don't know what I'm doing, and multiplexing network and sound IO on the same thread could block the audio event loop for too long. Though, as I see it - it shouldn't.

Running multiple threads is of course as option, however then the syncronization would become a bit problematic - overall I want sending the data to the network to not lag too much behind the audio capture. You could say I want "low-latency" there, but since I'm not an expret I don't really know how much latency is considered low latency.

**Comment by knappador** [comment]
> My use case is I just want to grab the audio input and send it over the network

I think it's a bit extreme to introduce tokio dependence.  Writing the part to read Cpal and write tokio and reverse is more the domain of the user here.

**Comment by MOZGIII** [comment]
Ideally, if `cpal` claims to be low-level and being a library (i.e. not a framework), it should avoid forcing it's own eventloop implementation altogether, and just provide the primitives that user can call however he desires.
This is kind of a issue with pulse audio - it too bundles it's own eventloop. It's all good until you want to actually do the low level resource accounting, thread management and synchronization - a thing that I'd expect to be supported from a low-level library.

I feel it's contradicting the current state of things. Currently `cpal` is more like an easy-to-use abstraction, but it's definitely not as close-to-metal nor flexible as I'd imagine.

However, I'm not saying that `cpal` should actually do all this - it's definitely good as it is.

So, instead of building `cpal`'s event loop atop of `tokio` it could be a good idea to just provide a `tokio` compatibility layer via `future`s. It's not obvious how to implement cooperation correctly, so maybe it's a good idea for `cpal` to address that issue and provide some kind of a solution - it seems like it can be implemented in a more optimized way if it's coupled with the internals of the event-loop and `cpal` synchronization management.

**Comment by MOZGIII** [comment]
By the way, it's not a good idea to use `mio`-style event loop if the underlying system primitives does not provide enough support for polling via select/epoll.

**Comment by tomaka** [comment]
What you're asking may not be possible for MacOS and Android. As for Linux and Windows, it is essentially a rewrite of half of the code.
This may sound aggressive, but I don't know how to reply to this other than "pull requests welcome".

**Comment by MOZGIII** [comment]
That sounds reasonable, although I was hoping to have a discussion at least. I'm definitely not experienced enough in the audio area to fully understand the scope of the problem. So, yeah, doing more research is what I'll probably do.
As an addition - I tried implementing `futures` with a naive approach - i.e. without really understanding how they work, and how `cpal`'s supposed to work - and that (as I expected) didn't work great.
I'll definitely be looking into it more, but for now I wanted to know
a) if anyone knows how to do it already, as for somebody more experienced it may be just obvious;
and b) if, from the `cpal` architecture standpoint, the support for `futures`/`tokio` should be implemented as a part of (i.e. inside of) the `cpal` code base, or is it possible to implement it atop of the public interface of `cpal` (as a third party crate).
I know that I myself might've been rude - sorry about that - so here's just something (hopefully) more concrete to kick off the constructive discussion.

**Comment by tomaka** [comment]
It's definitely feasible to rewrite the Windows and Linux backends to use tokio, considering that we're polling respectively handles and descriptors. However, as I said, it's a huge amount of efforts.

It's also possible to implement futures on top of cpal by using a background thread and channels, but doing so wouldn't make much sense in terms of latency.

**Comment by MOZGIII** [comment]
This is very valuable input. Could you please elaborate more about latency? What's the problem there? Won't futures still be cost-free abstractions in the case of cpal?

**Comment by MOZGIII** [comment]
Oh, I somehow missed the part about background thread :/ That makes sense.

**Comment by ishitatsuyuki** [comment]
I believe audio processing should not happen in a general purpose event loop. Audio is latency sensitive, even more than rendering, and processing should happen on a high priority dedicated thread.

Therefore, I vote to close this as wontfix.

**Comment by tomaka** [comment]
I don't see your point. Which thread the event loop runs on is totally orthogonal to the fact that we expose the event loop. Right now it exists but it's hidden, this issue would make it more visible. Furthermore, prioritizing tasks using a non-default tokio executor is also an option.

**Comment by MOZGIII** [comment]
I agree with the latency requirements, however I don't think the library can manage high priority dedicated threads better than a programer can in a case-by-case basis.

I think at some point in the future I'll try implementing futures atop of cpal to integrate it with tokio for purposes of loading the data in and from the cpal event loop. I tried that before and failed, mainly becuase I was at the very early stages of learning rust. Now I think I've learned enough to try tackling this again.
